File: go/code_smells/oop_abusers/temporal_instance_variables.go
package oop_abusers

import (
	"fmt"
	"strings"
)

// Code smell: Temporal Instance Variables [Variables de instancia temporales].
// PizzaOrder utiliza variables de instancia que solo son válidas entre la llamada
// a Start() y Place().

// Ejercicio: Añade una validación para que no se pueda llamar a Place()
// si no se ha añadido al menos un ingrediente.

// Te darás cuenta de que el objeto es una "máquina de estados" frágil.
// Si un cliente olvida llamar a Start() o intenta llamar a AddTopping() fuera de orden,
// el sistema puede fallar silenciosamente o requerir comprobaciones constantes
// de nulidad en cada método.

type PizzaOrder struct {
	size     *string
	toppings []string
	address  *string
}

func NewPizzaOrder() *PizzaOrder {
	return &PizzaOrder{
		toppings: []string{},
	}
}

func (p *PizzaOrder) Start(size string) {
	p.size = &size
	p.toppings = []string{}
	p.address = nil
}

func (p *PizzaOrder) AddTopping(topping string) {
	if p.size == nil {
		return
	}
	p.toppings = append(p.toppings, topping)
}

func (p *PizzaOrder) SetDeliveryAddress(address string) {
	p.address = &address
}

func (p *PizzaOrder) Place() string {
	sizeStr := "?"
	if p.size != nil {
		sizeStr = *p.size
	}
	addressStr := "UNKNOWN"
	if p.address != nil {
		addressStr = *p.address
	}
	summary := fmt.Sprintf("Pizza %s to %s with [%s]", sizeStr, addressStr, strings.Join(p.toppings, ", "))
	p.size = nil
	p.address = nil
	p.toppings = []string{}
	return summary
}

func DemoPizzaOrder() string {
	o := NewPizzaOrder()
	o.Start("L")
	o.AddTopping("pepperoni")
	o.AddTopping("mushroom")
	o.SetDeliveryAddress("123 Main St")
	return o.Place()
}

File: python/src/code_smells/oop_abusers/temporal_instance_variables.py
"""
Temporal Instance Variables Code Smell Example

Instance variables that are only valid during certain phases of an object's lifecycle.
This creates temporal coupling and makes the object's state inconsistent.

Refactoring suggestion: Replace Method with Method Object, Extract Class
"""
from typing import Literal


class PizzaOrder:
    """
    Instance variables (size, toppings, address) are only meaningful
    between start() and place() calls.

    The object goes through phases: uninitialized -> started -> placed -> uninitialized
    This temporal coupling makes the class fragile and error-prone.
    """

    def __init__(self) -> None:
        self._size: Literal['S', 'M', 'L'] | None = None
        self._toppings: list[str] = []
        self._address: str | None = None

    def start(self, size: Literal['S', 'M', 'L']) -> None:
        """Phase 1: Initialize order"""
        self._size = size
        self._toppings = []
        self._address = None

    def add_topping(self, topping: str) -> None:
        """Phase 2: Build order (only valid after start())"""
        if not self._size:
            return
        self._toppings.append(topping)

    def set_delivery_address(self, address: str) -> None:
        """Phase 2: Build order (only valid after start())"""
        self._address = address

    def place(self) -> str:
        """Phase 3: Finalize and reset"""
        summary = f"Pizza {self._size or '?'} to {self._address or 'UNKNOWN'} with [{', '.join(self._toppings)}]"
        # Reset temporal state - object returns to uninitialized phase
        self._size = None
        self._address = None
        self._toppings = []
        return summary


def demo_pizza_order() -> str:
    order = PizzaOrder()
    order.start('L')
    order.add_topping('pepperoni')
    order.add_topping('mushroom')
    order.set_delivery_address('123 Main St')
    return order.place()
